### PR TITLE
pivotal-157597056: pending submission submit api

### PR DESCRIPTION
**Pull request by olgamelnichuk**
Added direct submit call to the backend's REST API

### CODE REVIEW

File: BackendWebApp/src/test/java/uk/ac/ebi/biostd/webapp/application/rest/service/PendingSubmissionServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                .thenAnswer((Answer<UserData>) invocation -> {
+                    UserData ud = new UserData();
+                    ud.setData(invocation.getArgument(2));
+                    return ud;
+                });
+
+        Optional<PendingSubmissionDto> result = pendingSubmissionService.createSubmission(json, user);
+        assertThat(result.isPresent()).isTrue();
+        result.ifPresent(
+                r -> {
```
Comment (Juan-EBI) [inline BackendWebApp/src/test/java/uk/ac/ebi/biostd/webapp/application/rest/service/PendingSubmissionServiceTest.java]: I do not link that much this veroboty of using lamda when we already assert result optional is present, I would suggest to declared optional variable as `optioanlResult` and add new variable `result`

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/legacy/persistence/EntityManagerFilter.java [hunk 2]
```diff
@@ -18,7 +19,10 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
         try {
             chain.doFilter(req, res);
         } finally {
-            BackendConfig.getServiceManager().closeEntityManager();
+            ServiceManager sm = BackendConfig.getServiceManager();
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/legacy/persistence/EntityManagerFilter.java]: Yep, but what if ServiceManager is null? 
mockmvc tests are failing with NPE error in the filter.  Any other ideas?

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/ApiErrorDto.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package uk.ac.ebi.biostd.webapp.application.rest.dto;
+
+import lombok.Data;
+
+@Data
+public class ApiErrorDto {
+
+    private String message;
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/ApiErrorDto.java]: Sure

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/LogNodeDto.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    @Builder.Default
+    private String message = "";
+
+    private List<LogNodeDto> subnodes;
+
+    public static LogNodeDto from(LogNode log) {
+        return LogNodeDto.builder()
+                .level(log.getLevel())
+                .message(log.getMessage())
+                .subnodes(Optional.ofNullable(log.getSubNodes()).orElse(Collections.emptyList()).stream()
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/LogNodeDto.java]: Done. Thanks!

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/SectionMappingDto.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.Data;
+import uk.ac.ebi.biostd.in.AccessionMapping;
+
+@Builder
+@Data
+public class SectionMappingDto {
+
+    private List<Integer> order;
+
+    @Builder.Default
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/SectionMappingDto.java]: It seems to be a mapping between original section accessions and actual ones. This mapping appears in the report log from the submit code. It looks to me like a map with an order.. hm?

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/SubmissionMappingDto.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    private String assigned = "";
+
+    private List<SectionMappingDto> sections;
+
+    public static SubmissionMappingDto from(SubmissionMapping mapping) {
+        return SubmissionMappingDto.builder()
+                .order(mapping.getSubmissionMapping().getPosition()[0])
+                .original(mapping.getSubmissionMapping().getOrigAcc())
+                .assigned(mapping.getSubmissionMapping().getAssignedAcc())
+                .sections(Optional.ofNullable(mapping.getSectionsMapping()).orElse(Collections.emptyList()).stream()
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/SubmissionMappingDto.java]: Sure. Done.

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/SubmitReportDto.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+public class SubmitReportDto {
+    private SubmitStatus status;
+    private List<SubmissionMappingDto> mapping;
+    private LogNodeDto log;
+
+    public static SubmitReportDto fromSubmissionReport(SubmissionReport report) {
+        return SubmitReportDto.builder()
+                .status(submitStatus(report.getLog().getLevel().getPriority() < LogNode.Level.ERROR.getPriority()))
+                .mapping(
+                        Optional.ofNullable(report.getMappings()).orElse(Collections.emptyList()).stream()
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/dto/SubmitReportDto.java]: Done

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/service/PendingSubmissionService.java [hunk 8]
```diff
@@ -39,16 +37,17 @@
 
     private final UserDataService userDataService;
     private final PendingSubmissionUtil util;
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/service/PendingSubmissionService.java]: :-)) sure. done. it's pendingSubmissionUtil now

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/service/PendingSubmissionService.java [hunk 9]
```diff
@@ -59,26 +58,60 @@ public PendingSubmissionListDto getSubmissionList(PendingSubmissionListFiltersDt
...
+
+        SubmitReportDto report = submitService.submitJson(data, operation, user);
+        if (report.getStatus() == SubmitStatus.OK) {
+            deleteSubmissionByAccNo(dto.getAccno(), user);
+        }
+        return report;
+    }
+
+    private JsonNode amendAccno(JsonNode pageTab) {
+        return new PageTabProxy(pageTab).setAccno(BSST_ACCNO_TEMPLATE).json();
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/service/PendingSubmissionService.java]: Thanks, I forgot to remove it. it doesn't need to be moved..

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/service/SubmitService.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        Optional<DataFormat> format = DataFormat.fromFileNameOrContentType(
+                file.getOriginalFilename(),
+                file.getContentType());
+        if (!format.isPresent()) {
+            return fromErrorMessage(format("Unrecognized data format: %s, %s", file.getOriginalFilename(), file.getContentType()));
+        }
+
+        byte[] bytes;
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/service/SubmitService.java]: Not sure I know how can I return an error log with SneakyThrows?

File: BackendWebApp/src/test/java/uk/ac/ebi/biostd/webapp/application/rest/controllers/SubmissionResourceTest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import uk.ac.ebi.biostd.webapp.application.rest.dto.SubmitReportDto;
+import uk.ac.ebi.biostd.webapp.application.rest.dto.SubmitStatus;
+import uk.ac.ebi.biostd.webapp.application.rest.mappers.SubmissionsMapper;
+import uk.ac.ebi.biostd.webapp.application.rest.service.SubmitService;
+
+@RunWith(SpringRunner.class)
+@WebMvcTest(SubmissionResource.class)
+@AutoConfigureMockMvc(secure = false)
+public class SubmissionResourceTest {
+    @Autowired
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/test/java/uk/ac/ebi/biostd/webapp/application/rest/controllers/SubmissionResourceTest.java]: Done

File: BackendWebApp/src/test/java/uk/ac/ebi/biostd/webapp/application/rest/service/PendingSubmissionServiceTest.java (hunk 1 above)
Comment (olgamelnichuk) [reply BackendWebApp/src/test/java/uk/ac/ebi/biostd/webapp/application/rest/service/PendingSubmissionServiceTest.java]: :-) Okay. Done.

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/legacy/persistence/EntityManagerFilter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Juan-EBI) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/legacy/persistence/EntityManagerFilter.java]: ahhh you will either remove this filter bean from test context for example declaring as a `MockBean`  or creating a test profile or  make a call to `BackendConfig#setServiceManager` in test setup

[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/legacy/persistence/EntityManagerFilter.java]: Found a better way.. @AutoConfigureMockMvc(secure = false, addFilters = false)

[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/legacy/persistence/EntityManagerFilter.java]: All done

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/controllers/PendingSubmissionResource.java [hunk 12]
```diff
@@ -20,38 +21,55 @@
 
     @GetMapping("/submissions/pending")
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/controllers/PendingSubmissionResource.java]: Done.

File: BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/controllers/SubmissionResource.java [hunk 13]
```diff
@@ -18,11 +25,36 @@
...
+            @AuthenticationPrincipal User user) {
+        return submitService.createOrUpdateSubmission(file, attachTo, accnoTemplate, operation, user);
+    }
+
+    @InitBinder
+    public void initBinder(final WebDataBinder webdataBinder) {
+        webdataBinder.registerCustomEditor(SubmitOperation.class, new PropertyEditorSupport() {
+            @Override
+            public void setAsText(String text) throws IllegalArgumentException {
+                SubmitOperation value = Arrays.stream(SubmitOperation.values())
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply BackendWebApp/src/main/java/uk/ac/ebi/biostd/webapp/application/rest/controllers/SubmissionResource.java]: Done

**Pull request merged**